### ISSUE TITLE
Array constructor copy size seems wrong

### ISSUE THREAD

**Issue by demoitem** [modula3/cm3#264]
This small program extracted from p277 VArrsInArr causes a segv using the llvm backend, yet the m3cc backend seems
to work for unknown reason. Stepping through the code in assembly causes gdb to complain about the stack being smashed or corrupted.

I think the size of the copy is too large for each array element being initialised.

MODULE Main ;
IMPORT IO;
TYPE 
  Rng = [0 .. 31];
  Arr = ARRAY [0 .. 3] OF BITS 5 FOR Rng;

CONST A1C = Arr{1, 4, 7, 10};
VAR A1V := A1C;
CONST A2C = Arr{9, 6, 3, 0};
VAR A2V := A2C;
CONST A3C = Arr{2, 5, 8, 11};
VAR A3V := A3C;

VAR
  return : INTEGER;

PROCEDURE V () : INTEGER =
  VAR 
    Loc := ARRAY [1 .. 3] OF Arr{A1V , A2V , A3V};   <------- problem here
  BEGIN
(*
    Loc := ARRAY [1 .. 3] OF Arr{A1V, A2V, A3V};
*)
    RETURN 0;
  END V;

BEGIN
  return := V();
END Main.

This is the generated M3 IR

	begin_unit	 0
	widechar_size	 16
					-----FILE ../src/Main.m3  -----
					-----LINE 4  -----
	# module global constants
	declare_segment	 * -1 T v.1
	# module global data
	declare_segment	 M_Main -1 F v.2
					-----LINE 32  -----
	declare_procedure	 Main_M3 1 Addr 0 0 T * p.1
	declare_param	 mode 8 8 Int.64 425470580 F F 100 v.3
	declare_subrange	 765876253 425470580 0 31 8
	declare_subrange	 -1993854505 425470580 0 3 8
	declare_packed	 2086536086 5 765876253
	declare_array	 1916360575 -1993854505 2086536086 32
	declare_proctype	 -291811747 0 425470580 0 0
	declare_subrange	 -1993854761 425470580 1 3 8
	declare_array	 365186788 -1993854761 1916360575 96
	import_procedure	 Main_I3 0 Addr 0 p.2
	export_unit	 Main
	import_procedure	 IO_I3 0 Addr 0 p.3
	import_unit	 IO
	import_procedure	 RTHooks_I3 0 Addr 0 p.4
	import_unit	 RTHooks
	declare_typename	 765876253 Rng
	declare_typename	 1916360575 Arr
	declare_indirect	 -1916360576 1916360575
	declare_procedure	 Main__V 0 Int.64 0 0 F * p.5
	declare_local	 Loc 12 4 Struct 365186788 T F 50 v.4
	declare_local	 _result 8 8 Int.64 425470580 F F 50 v.5
	# V
					-----LINE 22  -----
	begin_procedure	 p.5
					-----LINE 24  -----
	load_address	 v.4 0
	load_address	 v.2 112
	copy		 4 Int.32 F   <----------------------- I think this is copying 1 array elt, size 4 bytes but it's 4 times size of Int.32 which is 16 bytes - same for other 3 copies below. So maybe its either    1 Int.32 or 4 Int.8
	load_address	 v.4 4
	load_address	 v.2 128
	copy		 4 Int.32 F
	load_address	 v.4 8
	load_address	 v.2 144
	copy		 4 Int.32 F
	import_procedure	 alloca 1 Addr 0 p.6
	declare_param	 * 8 8 Word.64 0 F F 0 v.6
	import_global	 m3_jmpbuf_size 8 8 Word.64 0 v.7
	# AllocateJmpbufs V try_count:0
					-----LINE 23  -----
					-----LINE 29  -----
	load_integer	 Int.64 0
	exit_proc	 Int.64
	end_procedure	 p.5
	# Main_M3
	# module main body Main_M3
					-----LINE 32  -----
	begin_procedure	 p.1
	load		 v.3 0 Int.64 Int.64
	if_false	 Int.64 L.1 0
	# AllocateJmpbufs  try_count:0
					-----LINE 33  -----
	start_call_direct	 p.5 0 Int.64
	call_direct	 p.5 Int.64
	declare_temp	 8 8 Int.64 F v.8
	store		 v.8 0 Int.64 Int.64
	load		 v.8 0 Int.64 Int.64
	store		 v.2 152 Int.64 Int.64
. L.1 F
	load_address	 v.2 0
	exit_proc	 Addr
	free_temp	 v.8
	end_procedure	 p.1
	# global constant type descriptor
	declare_record	 -1 704 3
	declare_field	 A1C 0 32 1916360575
	declare_field	 A2C 32 32 1916360575
	declare_field	 A3C 64 32 1916360575
	# global data type descriptor
	declare_record	 -1 1856 7
	declare_field	 A1C 832 64 -1916360576
	declare_field	 A1V 896 32 1916360575
	declare_field	 A2C 960 64 -1916360576
	declare_field	 A2V 1024 32 1916360575
	declare_field	 A3C 1088 64 -1916360576
	declare_field	 A3V 1152 32 1916360575
	declare_field	 return 1216 64 425470580
	# module global constants
	bind_segment	 v.1 88 8 Struct F T
	begin_init	 v.1
	# Contents of constant Main.A1C
	init_int	 0 129 Int.8
	# Contents of constant Main.A2C
	init_int	 1 28 Int.8
	init_int	 2 5 Int.16
	init_int	 4 201 Int.8
	# Contents of constant Main.A3C
	init_int	 5 12 Int.8
	init_int	 8 162 Int.8
	# procedure names
	init_int	 9 160 Int.8
	init_int	 10 5 Int.16
	init_chars	 16 "Main_M3"
	init_chars	 24 "V"
	# procedure table
	init_proc	 32 p.1
	init_var	 40 v.1 16
	init_proc	 48 p.5
	init_var	 56 v.1 24
	# file name
	init_chars	 72 "../src/Main.m3"
	end_init	 v.1
	# module global data
	bind_segment	 v.2 232 8 Struct F T
	begin_init	 v.2
	init_var	 0 v.1 72
	init_var	 40 v.1 32
	init_var	 72 v.2 160
	init_proc	 88 p.1
	init_int	 96 3 Int.64
	# Address of constant Main.A1C
	init_var	 104 v.1 0
	init_int	 112 129 Int.8
	# Address of constant Main.A2C
	init_int	 113 28 Int.8
	init_int	 114 5 Int.16
	init_var	 120 v.1 4
	init_int	 128 201 Int.8
	# Address of constant Main.A3C
	init_int	 129 12 Int.8
	init_var	 136 v.1 8
	init_int	 144 162 Int.8
	init_int	 145 160 Int.8
	init_int	 146 5 Int.16
	init_proc	 168 p.2
	init_var	 176 v.2 184
	init_proc	 192 p.3
	init_var	 200 v.2 208
	init_proc	 216 p.4
	end_init	 v.2
	# load map
	# 
	# 
	#  global data allocation for M_Main
	#      0   104  8  *module info*
	#    104     8  8  constantMain.A1C_ADDR_
	#    112     4  4  Main.A1V
	#    120     8  8  constantMain.A2C_ADDR_
	#    128     4  4  Main.A2V
	#    136     8  8  constantMain.A3C_ADDR_
	#    144     4  4  Main.A3V
	#    152     8  8  Main.return
	#    160    24  8  import Main
	#    184    24  8  import IO
	#    208    24  8  import RTHooks
	#    232     0  8  *TOTAL*
	# 
	# 
	#  global constants for M_Main
	#      0     4  4  constant Main.A1C
	#      4     4  4  constant Main.A2C
	#      8     4  4  constant Main.A3C
	#     16    10  8  *proc names*
	#     32    40  8  *proc info*
	#     72    15  1  *string*
	#     88     0  8  *TOTAL*
	# 
	end_unit

**Comment by RodneyBates** [comment]
Yes, your are right.  It's copying too much.  I guess it appears to work in some cases
because the elements of the outer array are being copied left to right, so some of
the excess garbage is overlaid by the following copy, making  Loc itself correct,
but stepping on other stuff.

This is why I give test cases random, varied, odd, etc. numbers, to reduce the chance
of accidental or coincidental correct working, but it didn't catch this one.

**Comment by RodneyBates** [comment]
On 3/29/21 1:34 AM, peter mckinna wrote:
> This small program extracted from p277 VArrsInArr causes a segv using the llvm backend, yet the m3cc backend seems
> to work for unknown reason. Stepping through the code in assembly causes gdb to complain about the stack being smashed or corrupted.
> 
> I think the size of the copy is too large for each array element being initialised.

Yes, the length is too large.  I have a simple one-line fix.  Before I tried it out,
I tried to put canaries in this test case to prove it's wrong and look for that to
go away afterwards.  I finally gave up.  I can't simultaneously git the compiler to
put the canary where it needs to be and initialize it before the errant code executes.

In the process, I found in some cases, at least, that it is copying into a stack temp,
and these are all beyond any programmer-declared variables.  So that means it was
writing into space beyond top of stack.

Also, I can't compile a compiler right now, until I can manually clean up file damage
resulting from a disk partition filling up and not noticed right away.  So unless it
is blocking you or somebody on something, I'll just leave it for now.  Otherwise, I
can post the patch, but it has had no testing.

**Comment by demoitem** [comment]
No worries.

I'm not blocked on anything so take your time.
